lib/cc-cubic.c

```diff
@@ -0,0 +1,155 @@
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include <math.h>
+#include "quicly/cc.h"
+
+#define QUICLY_MIN_CWND 2
+
+typedef double cubic_float_t;
+#define QUICLY_CUBIC_C ((cubic_float_t)0.4)
+#define QUICLY_CUBIC_BETA ((cubic_float_t)0.7)
+
+/* Calculates the elapsed time since the last congestion event (parameter t) */
+static cubic_float_t calc_cubic_t(const quicly_cc_t *cc)
+{
+    cubic_float_t clock_delta = clock() - cc->state.cubic.avoidance_start;
+    return clock_delta / CLOCKS_PER_SEC;
+}
+
+/* Equation 1 (using bytes as unit instead of MSS) */
+static uint32_t calc_w_cubic(const quicly_cc_t *cc, cubic_float_t t_sec, uint32_t max_udp_payload_size)
+{
+    cubic_float_t tk = t_sec - cc->state.cubic.k;
+    return (QUICLY_CUBIC_C * (tk * tk * tk) * max_udp_payload_size) + cc->state.cubic.w_max;
+}
+
+/* Equation 2 */
+/* K depends solely on W_max, so we update both together on congestion events */
+static void update_cubic_k(quicly_cc_t *cc, uint32_t max_udp_payload_size)
+{
+    cubic_float_t w_max_mss = cc->state.cubic.w_max / (cubic_float_t)max_udp_payload_size;
+    cc->state.cubic.k = cbrt(w_max_mss * ((1 - QUICLY_CUBIC_BETA) / QUICLY_CUBIC_C));
+}
+
+/* Equation 4 (using bytes as unit instead of MSS) */
+static uint32_t calc_w_est(const quicly_cc_t *cc, cubic_float_t t_sec, cubic_float_t rtt_sec, uint32_t max_udp_payload_size)
+{
+    return (cc->state.cubic.w_max * QUICLY_CUBIC_BETA) +
+           ((3 * (1 - QUICLY_CUBIC_BETA) / (1 + QUICLY_CUBIC_BETA)) * (t_sec / rtt_sec) * max_udp_payload_size);
+}
+
+static void quicly_cubic_init(quicly_cc_t *cc, const quicly_cc_conf_t *conf, uint32_t initcwnd)
+{
+    cc->cwnd = cc->cwnd_initial = cc->cwnd_maximum = initcwnd;
+    cc->ssthresh = cc->cwnd_minimum = UINT32_MAX;
+}
+
+/* TODO: Avoid increase if sender was application limited. */
+static void quicly_cubic_on_acked(quicly_cc_t *cc, const quicly_loss_t *loss, uint32_t bytes, uint64_t largest_acked,
+                                  uint32_t inflight, uint32_t max_udp_payload_size)
```

Comment:
We can remove `quicly_` from static functions being defined, as they do not pollute the global namespace. We can remove `cubic_` too, though I am fine with retaining it as it might turn out to be helpful when we look at stack traces.

Reply:
Left `cubic_`/`reno_` in for now. Makes searching easier as well.